### PR TITLE
docs: deprecate ViewEngine-based JIT Compiler symbols

### PR DESCRIPTION
**Pull request by AndrewKushnir**
**NOTE: JIT compiler in Ivy is available, these deprecations are for ViewEngine only.**

DEPRECATION:

Some of the symbols that are used in ViewEngine-based JIT Compiler API are now deprecated:

- `ModuleWithComponentFactories`
- `Compiler`
- `CompilerFactory`
- `JitCompilerFactory`
- `NgModuleFactory`

In Ivy JIT mode those symbols are unused and they are currently present for backwards-compatibility with ViewEngine only. Since ViewEngine is deprecated and will shortly be fully removed, those symbols are now deprecated as well.

## PR Type
What kind of change does this PR introduce?

- [x] Documentation content changes

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by AndrewKushnir** [comment]
@IgorMinar @alxhub @JoostK could you please have a look at this draft PR and let me know if you have any feedback? Thank you.

File: packages/core/src/linker/compiler.ts [hunk 1]
```diff
@@ -26,6 +26,11 @@ import {NgModuleFactory} from './ng_module_factory';
  * Combination of NgModuleFactory and ComponentFactories.
  *
  * @publicApi
+ *
+ * @deprecated
+ * This class was used in ViewEngine-based JIT API (see `Compiler` class) and is no
+ * longer needed in Ivy JIT mode. Since ViewEngine is deprecated and will shortly be fully removed,
+ * the `ModuleWithComponentFactories` class is now deprecated as well.
```
Comment (JoostK) [inline packages/core/src/linker/compiler.ts:33]: This type did provide access to all components in an NgModule, which isn't available without this API even in Ivy.

**Review by JoostK — CHANGES_REQUESTED** [review]
While it is true that Ivy's decorators drive the JIT compiler instead of those classes, the `Compiler` service has been the one and only source for getting `NgModuleFactory` symbols, even outside of JIT usages. So I would personally avoid phrasing this as a fully unnecessary API in Ivy, as it does have its use-cases.

I believe that `createNgModuleRef` alleviates the need for creating `NgModuleFactory` instances and we should probably point people towards using that API. We don't provide an alternative for `ModuleWithComponentFactories` and that API is available in Ivy as well, so I'm not sure what our messaging should be here.

tldr; I think we shouldn't make it seem like these APIs are ViewEngine specific and irrelevant in Ivy (throughout this PR)

**Review by alxhub — APPROVED** [review]

File: packages/core/src/linker/compiler.ts (hunk 1 above)
Comment (IgorMinar) [reply packages/core/src/linker/compiler.ts:33]: you are right. the only reason why someone would use this is if there were doing something very dynamic using metaprograming. I've heard of a few trying such thing, but I don't know if is actually used in production anywhere.

I'm ok with deprecating this now, and soliciting feedback via issues that would then inform a more targeted replacement API - *if* someone asks for it.

**Review by IgorMinar — APPROVED** [review]
LGTM overall, but I want to +1 @JoostK comments and also ask that we tweak the wording so that it's clear from the deprecation page that JIT is not deprecated, only some APIs used mostly with JIT and only with VE are going away.

Reviewed-for: public-api, docs-packaging-and-releasing

**Review by atscott — APPROVED** [review]
reviewed-for: public-api

**Review by alxhub — APPROVED** [review]
Reviewed-for: public-api

**Review by IgorMinar — APPROVED** [review]
Looks great. Thanks

**Comment by dylhunn** [comment]
Force merging as requested.

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit e3d6e97050ef7637b7dd80d5c018a24cabb9caee.
